dcpquery/db/__init__.py

```diff
@@ -174,8 +136,126 @@ def create_recursive_process_functions_in_db():
                             AND file_uuid=NEW.file_uuid
                         )
                         DO INSTEAD NOTHING;
-                    """)
-    config.db_session.commit()
+        """
+
+
+class ProcessProcessLink(SQLAlchemyBase):
+    __tablename__ = 'process_join_table'
+    id = Column(Integer, primary_key=True)
+    child_process_uuid = Column(UUID, ForeignKey("processes.process_uuid"))
+    parent_process_uuid = Column(UUID, ForeignKey("processes.process_uuid"))
+    parent_process = relationship(Process, foreign_keys=[parent_process_uuid])
+    child_process = relationship(Process, foreign_keys=[child_process_uuid])
+
+
+class InitDB:
+    def __init__(self, dry_run=True):
+        from sqlalchemy_utils import database_exists, create_database
+
+        logger.info("Initializing database at %s", repr(config.db.url))
+        if not database_exists(config.db.url):
+            logger.info("Creating database")
+            create_database(config.db.url)
+        logger.info("Initializing database")
+
+        if dry_run:
+            config._db_engine_params.update(strategy="mock", executor=lambda sql, *args, **kwargs: print(sql))
+        SQLAlchemyBase.metadata.create_all(config.db)
+        self.create_recursive_functions_in_db()
+        self.create_upsert_rules_in_db()
+
+    upsert_process_file_link_rule_sql = """
+                        CREATE OR REPLACE RULE process_file_join_table_ignore_duplicate_inserts AS
+                            ON INSERT TO process_file_join_table
+                                WHERE EXISTS (
+                                  SELECT 1
+                                FROM process_file_join_table
+                                WHERE process_uuid = NEW.process_uuid
+                                AND process_file_connection_type=NEW.process_file_connection_type
+                                AND file_uuid=NEW.file_uuid
+                            )
+                            DO INSTEAD NOTHING;
+                        """
```

Comment:
Please format the closing quotes to be at a consistent indent. My preference is for them to be at the same indent as the variable assignment line (line 167 here, 4 spaces).

Similarly, please format the SQL to be at a fixed indent. My preference is to indent it one level in from the variable assignment line (so here, line 180 would be indented 8 spaces).

Reply:
👍